### ISSUE TITLE
Error starting PowerShell 7.4.0

### ISSUE THREAD

**Issue by giggio** [PowerShell/PowerShell#20706]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

1. Install PowerShell 7.4.0
2. Start it
3. See it fail

### Expected behavior

```console
It starts, like 7.3 did.
```

### Actual behavior

```console
It fails with:


PowerShell 7.4.0
The shell cannot be started. A failure occurred during initialization:
Index was out of range. Must be non-negative and less than or equal to the size of the collection. (Parameter 'startIndex')
```
```


### Error details

```console
Not possible.
```

### Environment data

```powershell
Not possible, but it is version 7.4.0.
```

### Visuals

This is starting from Nushell:

![from nu](https://github.com/PowerShell/PowerShell/assets/334958/1ea4f9f3-686a-4134-a945-d7d83c8edb42)

This is starting it from Windows Terminal:

![from wt](https://github.com/PowerShell/PowerShell/assets/334958/b35967ca-eaea-45d7-9648-e5abcf9e89a8)

Downloaded and installed this file: https://github.com/PowerShell/PowerShell/releases/download/v7.4.0/PowerShell-7.4.0-win-x64.msi

**Comment by rhubarb-geek-nz** [comment]
Can you check with "Add & Remove" settings that there is only one PowerShell MSI installed, and you don't have both 7.3.9 and 7.4.0?

**Comment by giggio** [comment]
@rhubarb-geek-nz I did, I hadn't.

**Comment by B-Art** [comment]
There are at least 4 ways to install Pwsh on Windows:
From MSStore, via MSI, via Winget and Unzipping the download.
Which one did you use @giggio ?

**Comment by giggio** [comment]
First I tried via winget. Then via the MSI download. Same result.

**Comment by rhubarb-geek-nz** [comment]
> First I tried via winget. Then via the MSI download. Same result.

Have you tried completely uninstalling the previous MSI first? Not ideal I know, in-place updates should work.

**Comment by giggio** [comment]
> Have you tried completely uninstalling the previous MSI first? Not ideal I know, in-place updates should work. Use "Add or Remove programs" to see what is installed, and also check the directory "c:\Program Files\PowerShell\7"

I have.

I have now also tried installing the Microsoft Store Version, and I get the same error:
![image](https://github.com/PowerShell/PowerShell/assets/334958/9f51dd2f-710c-48fa-9332-4970cd912aae)

**Comment by giggio** [comment]
Tried now using the global tool, it also does not work:

![image](https://github.com/PowerShell/PowerShell/assets/334958/3d71bb41-2ec6-4598-93e3-e3d0bce093c0)

I installed `7.3.10`, as it still works. :(

**Comment by rhubarb-geek-nz** [comment]
Alas, there are enough other problems with 7.4.0 to suggest sticking with 7.3.10.

**Comment by giggio** [comment]
The problem persists at version `7.4.1`, released more recently.

![image](https://github.com/PowerShell/PowerShell/assets/334958/2f010937-215a-459a-8346-13a0b910ce76)

**Comment by giggio** [comment]
Also, tried using version 7.5 preview from the store, also broken.

![image](https://github.com/PowerShell/PowerShell/assets/334958/71dd6fb4-0c3d-441e-be1b-25fb934d987d)

**Comment by giggio** [comment]
I have tracked the error and added the explanation of why it is happening to the additional context section of the issue.

**Comment by w-boyd** [comment]
On Linux:
```
ForEach-Object: /home/vsts/work/_temp/f3a905a8-4984-469a-b3ba-7a9ba49e76af.ps1:5
Line |
   5 |  "whatever" | ForEach-Object -Parallel { $_ }
     |               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | An unexpected error has occurred while processing ForEach-Object
     | -Parallel input. This may mean that some of the piped input did not get
     | processed. Error: System.ArgumentOutOfRangeException: Index was out of
     | range. Must be non-negative and less than or equal to the size of the
     | collection. (Parameter 'startIndex')    at System.String.IndexOf(Char
     | value, Int32 startIndex, Int32 count)    at
     | System.Management.Automation.ModuleIntrinsics.UpdatePath(String path,
     | String pathToAdd, Int32& insertIndex)    at
     | System.Management.Automation.ModuleIntrinsics.GetModulePath(String
     | currentProcessModulePath, String hklmMachineModulePath, String
     | hkcuUserModulePath)    at
     | System.Management.Automation.ModuleIntrinsics.SetModulePath()    at
     | System.Management.Automation.ModuleIntrinsics..ctor(ExecutionContext
     | context)    at
     | System.Management.Automation.ExecutionContext.InitializeCommon(AutomationEngine engine, PSHost hostInterface)    at System.Management.Automation.AutomationEngine..ctor(PSHost hostInterface, InitialSessionState iss)    at System.Management.Automation.Runspaces.LocalRunspace.DoOpenHelper()    at System.Management.Automation.Runspaces.RunspaceBase.CoreOpen(Boolean syncCall)    at System.Management.Automation.PSTasks.PSTaskPool.GetRunspace(Int32 taskId)    at System.Management.Automation.PSTasks.PSTaskPool.Add(PSTaskBase task)    at Microsoft.PowerShell.Commands.ForEachObjectCommand.<InitParallelParameterSet>b__67_2(Object _).
```


On Windows:
```
"C:\Program Files\PowerShell\7\pwsh.exe" -NoLogo -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command ". 'D:\a\_temp\86498521-8914-4ea7-ba0c-67b8bb7a6633.ps1'"
The shell cannot be started. A failure occurred during initialization:
Index was out of range. Must be non-negative and less than or equal to the size of the collection. (Parameter 'startIndex')
##[error]PowerShell exited with code '70'.
```

Azure DevOps pipeline yaml to reproduce the issue:
```
jobs:
- job: windows
  pool:
      vmImage: windows-latest
  steps:
  - pwsh: |
      $filePath = Join-Path $env:SYSTEM_DEFAULTWORKINGDIRECTORY "powershellmodules"
      $separator = $isWindows ? ";" : ":"   
      $updatedPSModulePath = $env:PSModulePath, $filePath -join $separator
      $powershellConfigPath = Join-Path $PSHOME 'powershell.config.json'
      if (-not (Test-Path $powershellConfigPath)){
        New-Item -Path $powershellConfigPath -ItemType File -Force
        $config = @{
          PSModulePath = $updatedPSModulePath
        }
      }
      else {
        Write-Verbose "Found existing powershell config, updating PSModulePath"
        $config = Get-Content -Path $powershellConfigPath | ConvertFrom-Json -AsHashtable -Depth 100
        $config["PSModulePath"] = $updatedPSModulePath
      }
      Write-Verbose "Pushing content to $powershellConfigPath" -Verbose
      $config | ConvertTo-Json -Depth 100 | Tee-Object -FilePath $powershellConfigPath
    displayName: Update PSModulePath

  - pwsh: |
      Write-Verbose ('$env:PSModulePath {0}' -f $env:PSModulePath) -Verbose
      "whatever" | ForEach-Object -Parallel { $_ }
    displayName: Debug

- job: linux
  pool:
      vmImage: ubuntu-latest
  steps:
  - pwsh: |
      $filePath = Join-Path $env:SYSTEM_DEFAULTWORKINGDIRECTORY "powershellmodules"
      $separator = $isWindows ? ";" : ":"   
      $updatedPSModulePath = $env:PSModulePath, $filePath -join $separator
      $powershellConfigPath = Join-Path $PSHOME 'powershell.config.json'
      if (-not (Test-Path $powershellConfigPath)){
        New-Item -Path $powershellConfigPath -ItemType File -Force
        $config = @{
          PSModulePath = $updatedPSModulePath
        }
      }
      else {
        Write-Verbose "Found existing powershell config, updating PSModulePath"
        $config = Get-Content -Path $powershellConfigPath | ConvertFrom-Json -AsHashtable -Depth 100
        $config["PSModulePath"] = $updatedPSModulePath
      }
      Write-Verbose "Pushing content to $powershellConfigPath" -Verbose
      $config | ConvertTo-Json -Depth 100 | Tee-Object -FilePath $powershellConfigPath
    displayName: Update PSModulePath

  - pwsh: |
      Write-Verbose ('$env:PSModulePath {0}' -f $env:PSModulePath) -Verbose
      "whatever" | ForEach-Object -Parallel { $_ }
    displayName: Debug
```

**Comment by hoffmanntravis** [comment]
I have reproduced this issue on Powershell 7.4.6 locally.  This works:

```
{
    "PSModulePath": ":"
}
```

This does not:
```
{
    "PSModulePath": ";"
}
```

File Path:
C:\Users\<user>\Documents\PowerShell\powershell.config.json

This seems like a clear case of regression.  Please investigate and fix.

**Comment by hoffmanntravis** [comment]
Couple more notes.  This is still broken in Powershell Preview 7.5:
```
PowerShell 7.5.0-preview.5
The shell cannot be started. A failure occurred during initialization:
Index was out of range. Must be non-negative and less than or equal to the size of the collection. (Parameter 'startIndex')

[process exited with code 70 (0x00000046)]
You can now close this terminal with Ctrl+D, or press Enter to restart.
```

Also, note that in the example I listed in my last comment, the semicolon delimiter by itself alone is enough to break the startup of powershell.

According to this document, `;` should be the delimiting character:
https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_psmodulepath?view=powershell-7.4#modifying-psmodulepath

Editing the config file is described here:
https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_powershell_config?view=powershell-7.4#psmodulepath

**Comment by jborean93** [comment]
Opened https://github.com/PowerShell/PowerShell/pull/24568 to fix this problem.

**Comment by hoffmanntravis** [comment]
@jborean93 First, I just wanted to say thanks for putting a PR together and getting this some attention after a long time idle.

TLDR: The underlying thesis of my comments below are we are an impacted customer of fairly advanced users that are feeling a lot of pain from this, so maybe that tips the scales towards merging.  I read through the PR and noticed your comment below.

_"Waiting on an answer from the WG also takes way too much time and if I'm being honest I doubt this one will actually get merged anytime soon which is a pity.

Maybe I'll just bite the bullet and do it because I really dislike the current behavior and how it works."_

In response to that, I want to put out that we are a medium size azure customer that is fairly heavily impacted by this.  We use Azure DevOps with Azure.  We built our YAML deploy pipelines and our internal libraries around the notion of PSModulePath being overridable and extensible by deployment agents and users.  When we use AzurePowershell@5, it actually overrides any previously set environment variables unless we set one on the task in YAML directly.  This is burdensome compared to how it worked previously.  

Prior to the issue, we ran one task template with no parameters per job that writes the modules to disk, we update the machine environment variable, and we are done.  Developers can extend our libraries and templates with little to no knowledge of our tools.

Without this fix, it becomes very difficult to manage those modules and how we have it set up.  Yes, we can likely do an install-module or something, but checking out a repository, writing to disk, and setting the environment variable works great for us.  And then it broke so we've been stuck on 7.2 for months.  Recently I ended up writing my own randomizer and testing with NIST tools instead of using PS 7.4 Get-SecureRandom, which would have been so _much_ more convenient.

Not sure if that tips the scales or not in favor of merging but I hope it does 🙂.  And maybe provides a little cover if there are issues.  You can definitely blame me, because we're really feeling pain with out it, so there's one reason to do it anyway haha.  We really dislike the behavior right now ourselves, because it doesn't work for us at all and broke our existing process.

Thanks so much for your help!
Travis
